### PR TITLE
VReplication: Support automatically replacing auto_inc cols with sequences

### PR DESCRIPTION
**Pull request by mattlord**
## Description

Since https://github.com/vitessio/vitess/pull/15679 we automatically strip any [MySQL `auto_increment` clauses](https://dev.mysql.com/doc/refman/en/example-auto-increment.html) on tables that are being moved (using the [`MoveTables` command](https://vitess.io/docs/user-guides/migration/move-tables/)) from an unsharded keyspace to a sharded one. And since https://github.com/vitessio/vitess/pull/13656 we support initializing any sequences in the target keyspace on `SwitchTraffic`.

In this PR, we build upon both of those items of work to support automatically creating a [Vitess Sequence](https://vitess.io/docs/reference/features/vitess-sequences/) for each if those tables that HAD an `auto_increment` clause if one does not already exist. This requires the use of two flags:
  1. When creating the workflow, specify:
    -  That we want to replace any removed MySQL auto_increment clauses with vschema AutoIncrement definitions using `--sharded-auto-increment-handling=replace`
    - An unsharded keyspace in `--global-keyspace` that we can then later use to create the sequence tables in if they don't already exist (they don't in the manual test below)
  3. When switching traffic, specify the `--initialize-target-sequences` flag, which will create a missing sequence table when possible

### Manual test

```
alias vtctldclient='command vtctldclient --server=localhost:15999'

./101_initial_cluster.sh; mysql < ../common/insert_commerce_data.sql; ./201_customer_tablets.sh; ./202_move_tables.sh; ./203_switch_reads.sh; ./204_switch_writes.sh; ./205_clean_commerce.sh; ./301_customer_sharded.sh; ./302_new_shards.sh; ./303_reshard.sh; ./304_switch_reads.sh; ./305_switch_writes.sh; ./306_down_shard_0.sh; ./307_delete_shard_0.sh

vtctldclient ApplySchema --sql "alter table product drop primary key, add product_id int unsigned not null auto_increment primary key" commerce

vtctldclient ApplyVSchema --vschema='
{
  "sharded": true,
  "vindexes": {
    "hash": {
      "type": "hash"
    },
    "unicode_loose_xxhash": {
      "type": "unicode_loose_xxhash"
    }
  },
  "tables": {
    "corder": {
      "column_vindexes": [
        {
          "column": "customer_id",
          "name": "hash"
        }
      ],
      "auto_increment": {
        "column": "order_id",
        "sequence": "order_seq"
      }
    },
    "customer": {
      "column_vindexes": [
        {
          "column": "customer_id",
          "name": "hash"
        }
      ],
      "auto_increment": {
        "column": "customer_id",
        "sequence": "customer_seq"
      }
    },
    "product": {
      "column_vindexes": [
        {
          "column": "sku",
          "name": "unicode_loose_xxhash"
        }
      ]
    }
  }
}
' customer

sleep 5

vtctldclient MoveTables --workflow commerce2customer --target-keyspace customer create --source-keyspace commerce --tables product --sharded-auto-increment-handling=replace --global-keyspace commerce

vtctldclient GetVSchema customer --compact

mysql commerce -e "insert into product (sku, description, price) values ('SKU-1003', 'Mouse', 5)"

sleep 5

vtctldclient MoveTables --workflow commerce2customer --target-keyspace customer switchtraffic --initialize-target-sequences

vtctldclient GetVSchema customer --compact

mysql commerce -e "show tables"

mysql commerce -e "select * from product order by product_id"

mysql commerce -e "select * from product_seq"

mysql commerce -e "insert into product (sku, description, price) values ('SKU-1004', 'Table', 199), ('SKU-1005', 'Chair', 99)"

mysql commerce -e "select * from product order by product_id"

mysql commerce -e "select * from product_seq"
```

Results:
```
{
  "sharded": true,
  "vindexes": {
    "hash": {
      "type": "hash"
    }
  },
  "tables": {
    "corder": {
      "column_vindexes": [
        {
          "column": "customer_id",
          "name": "hash"
        }
      ],
      "auto_increment": {
        "column": "order_id",
        "sequence": "order_seq"
      }
    },
    "customer": {
      "column_vindexes": [
        {
          "column": "customer_id",
          "name": "hash"
        }
      ],
      "auto_increment": {
        "column": "customer_id",
        "sequence": "customer_seq"
      }
    },
    "product": {
      "column_vindexes": [
        {
          "column": "product_id",
          "name": "hash"
        }
      ],
      "auto_increment": {
        "column": "product_id",
        "sequence": "product_seq"
      }
    }
  }
}
+--------------------+
| Tables_in_commerce |
+--------------------+
| customer_seq       |
| order_seq          |
| product            |
| product_seq        |
+--------------------+
+----------+-------------+-------+------------+
+----------+-------------+-------+------------+
| sku      | description | price | product_id |
+----------+-------------+-------+------------+
| SKU-1001 | Monitor     |   100 |          1 |
| SKU-1002 | Keyboard    |    30 |          2 |
| SKU-1003 | Mouse       |     5 |          3 |
+----------+-------------+-------+------------+
+----+---------+-------+
| id | next_id | cache |
+----+---------+-------+
|  0 |       4 |  1000 |
+----+---------+-------+
+----------+-------------+-------+------------+
| sku      | description | price | product_id |
+----------+-------------+-------+------------+
| SKU-1001 | Monitor     |   100 |          1 |
| SKU-1002 | Keyboard    |    30 |          2 |
| SKU-1003 | Mouse       |     5 |          3 |
| SKU-1004 | Table       |   199 |          4 |
| SKU-1005 | Chair       |    99 |          5 |
+----------+-------------+-------+------------+
+----+---------+-------+
| id | next_id | cache |
+----+---------+-------+
|  0 |    1004 |  1000 |
+----+---------+-------+
```

The `strip_sharded_auto_increment` field was renamed to `sharded_auto_increment_handling` — which is upgrade/downgrade safe with protobufs as field indexes are used — and the type of the proto field was changed from `bool` to an `enum`/`int32` to support the new REPLACE option. This is upgrade/downgrade safe because [the wire format for bool is a `varint` with 0 being false and 1 being true](https://protobuf.dev/programming-guides/encoding/#structure) — with the deserialization ending up 0 = false > 0 = true. To demonstrate:
```
git checkout vrepl_replace_auto_inc && make build 
cp bin/vtctldclient /tmp

git checkout main && make build
cd examples/local
./101_initial_cluster.sh; mysql < ../common/insert_commerce_data.sql; ./201_customer_tablets.sh

❯ vtctldclient MoveTables --workflow commerce2customer --target-keyspace customer create --source-keyspace commerce --tables product --remove-sharded-auto-increment=REPLACE --global-keyspace commerce
E0930 17:51:16.473295   29269 main.go:56] invalid argument "REPLACE" for "--remove-sharded-auto-increment" flag: strconv.ParseBool: parsing "REPLACE": invalid syntax

❯ /tmp/vtctldclient --server=localhost:15999  MoveTables --workflow commerce2customer --target-keyspace customer create --source-keyspace commerce --tables product --sharded-auto-increment-handling=REPLACE --global-keyspace commerce
The following vreplication streams exist for workflow customer.commerce2customer:

id=1 on customer/zone1-200: Status: Copying. VStream has not started.

Traffic State: Reads Not Switched. Writes Not Switched

❯ command mysql -u root --socket ${VTDATAROOT}/vt_0000000$(vtctldclient GetTablets --keyspace customer --tablet-type primary --shard "0" | awk '{print $1}' | cut -d- -f2 | bc)/mysql.sock vt_customer -e "select * from _vt.vreplication\G"
*************************** 1. row ***************************
                   id: 1
             workflow: commerce2customer
               source: keyspace:"commerce" shard:"0" filter:{rules:{match:"product" filter:"select * from product"}}
                  pos: MySQL56/9fec34ea-7f75-11ef-8280-bf546605e365:1-45
             stop_pos: NULL
              max_tps: 0
  max_replication_lag: 0
                 cell: zone1
         tablet_types: replica,primary
         time_updated: 1727733136
transaction_timestamp: 0
                state: Running
              message:
              db_name: vt_customer
          rows_copied: 0
                 tags:
       time_heartbeat: 1727733136
        workflow_type: 1
       time_throttled: 0
  component_throttled:
     reason_throttled:
    workflow_sub_type: 0
 defer_secondary_keys: 0
              options: {"strip_sharded_auto_increment": true}
```

And using the manual test case from above but using binaries built on `main` with a `vtctldclient` binary from the PR branch:
```
...

❯ /tmp/vtctldclient --server :15999 MoveTables --workflow commerce2customer --target-keyspace customer create --source-keyspace commerce --tables product --sharded-auto-increment-handling=replace --global-keyspace commerce
The following vreplication streams exist for workflow customer.commerce2customer:

id=2 on customer/zone1-300: Status: Copying. VStream has not started.
id=2 on customer/zone1-401: Status: Copying. VStream has not started.

Traffic State: Reads Not Switched. Writes Not Switched

❯ mysql commerce/0 -e "show create table product\G"
*************************** 1. row ***************************
       Table: product
Create Table: CREATE TABLE `product` (
  `sku` varbinary(128) NOT NULL,
  `description` varbinary(128) DEFAULT NULL,
  `price` bigint DEFAULT NULL,
  `product_id` int unsigned NOT NULL AUTO_INCREMENT,
  PRIMARY KEY (`product_id`)
) ENGINE=InnoDB AUTO_INCREMENT=3 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci

❯ mysql customer/-80 -e "show create table product\G"
*************************** 1. row ***************************
       Table: product
Create Table: CREATE TABLE `product` (
  `sku` varbinary(128) NOT NULL,
  `description` varbinary(128) DEFAULT NULL,
  `price` bigint DEFAULT NULL,
  `product_id` int unsigned NOT NULL,
  PRIMARY KEY (`product_id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
```

So we can see that the auto_increment clauses were still removed.

## Related Issue(s)

  - Fixes: https://github.com/vitessio/vitess/issues/16864

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation: https://github.com/vitessio/website/pull/1859

### CODE REVIEW

File: go/vt/vtctl/workflow/traffic_switcher.go
Comment (rohit-nayak-ps) [inline go/vt/vtctl/workflow/traffic_switcher.go:1496]: This can be extracted into a separate function.

**Review by rohit-nayak-ps — APPROVED** [review]

File: go/cmd/vtctldclient/command/vreplication/movetables/create.go
Comment (mattlord) [reply go/cmd/vtctldclient/command/vreplication/movetables/create.go]: I believe that I've addressed these notes here: https://github.com/vitessio/vitess/pull/16860/commits/8d5087a0def7671c2992dde4e5b75fdcc2740bcf

File: go/cmd/vtctldclient/command/vreplication/movetables/movetables.go
Comment (mattlord) [reply go/cmd/vtctldclient/command/vreplication/movetables/movetables.go:112]: Addressed this here: https://github.com/vitessio/vitess/pull/16860/commits/8d5087a0def7671c2992dde4e5b75fdcc2740bcf

File: go/vt/vtctl/workflow/traffic_switcher.go
Comment (mattlord) [reply go/vt/vtctl/workflow/traffic_switcher.go:1496]: I moved it out here: https://github.com/vitessio/vitess/pull/16860/commits/eaed239217eae04f60f3734b0cf35ced179e7cd9

File: go/cmd/vtctldclient/command/vreplication/movetables/movetables.go
Comment (deepthi) [inline go/cmd/vtctldclient/command/vreplication/movetables/movetables.go:59]: Do we validate anywhere that this in fact an unsharded keyspace?

File: proto/vtctldata.proto
Comment (deepthi) [inline proto/vtctldata.proto:223]: We should rename this to `sharded_auto_increment_handling` or similar. Renaming proto fields does not break compatibility in anyway.

File: proto/vtctldata.proto
Comment (deepthi) [inline proto/vtctldata.proto:213]: WDYT of calling this `NONE` instead of `LEAVE`? Meaning we do nothing.

File: go/vt/vtctl/workflow/traffic_switcher_test.go
Comment (deepthi) [inline go/vt/vtctl/workflow/traffic_switcher_test.go:117]: there seems to be a mismatch between the field name and type. Should the field name also be `...Response`?

File: proto/vtctldata.proto
Comment (deepthi) [reply proto/vtctldata.proto:223]: Btw, when did we add this field? If it was within v21, we are fine. otherwise we'd have to reserve this and add a new field because we are changing the type.

File: go/cmd/vtctldclient/command/vreplication/common/utils.go
Comment (deepthi) [inline go/cmd/vtctldclient/command/vreplication/common/utils.go:290]: It's unclear exactly what the boolean logic here means to someone who doesn't already know it.
I'm reading it as
```
(IF the sequence table is not found) AND ((the sequence table reference was fully qualified) OR (a value was specified for --global-keyspace))
```
but it should be rewritten to avoid confusion.

File: go/cmd/vtctldclient/command/vreplication/movetables/create.go
Comment (deepthi) [inline go/cmd/vtctldclient/command/vreplication/movetables/create.go:44]: Along with the proto name change, we should change this name. It is confusing to call it `Strip` when it could be NONE/REMOVE/REPLACE.

Comment (deepthi) [inline go/cmd/vtctldclient/command/vreplication/movetables/create.go:44]:
```suggestion
		ShardedAutoIncrementHandling string
```

Comment (deepthi) [reply go/cmd/vtctldclient/command/vreplication/movetables/create.go:44]: There seems to be some nuance here that I missed during review. So maybe you just need to explain what I'm missing :)
Quoting Rohit:
> Maybe consider naming the StripShardedAutoIncrement differently for the string and enum representations. Also adding a comment to note that the first one is from the CLI flags and second to set in the _vt.vreplication row.

File: proto/vtctldata.proto
Comment (mattlord) [reply proto/vtctldata.proto:223]: Good point on renaming. Will do. That field was added in v20.

File: go/cmd/vtctldclient/command/vreplication/movetables/movetables.go
Comment (mattlord) [reply go/cmd/vtctldclient/command/vreplication/movetables/movetables.go:59]: Yes: https://github.com/vitessio/vitess/pull/16860/files#diff-cd06c8cb327b433a510fba2e8511819322fd13da0964396ddd54510c70580277R1508-R1520

We could also do that in MoveTablesCreate though.

File: proto/vtctldata.proto
Comment (mattlord) [reply proto/vtctldata.proto:213]: I'm fine either way. If you prefer that then I'm find changing it.

File: go/vt/vtctl/workflow/traffic_switcher_test.go
Comment (mattlord) [reply go/vt/vtctl/workflow/traffic_switcher_test.go:117]: It gets a big long, and this is test-only code where you can clearly see the type, but I can change it.

File: go/cmd/vtctldclient/command/vreplication/common/utils.go
Comment (mattlord) [reply go/cmd/vtctldclient/command/vreplication/common/utils.go:290]: I can think about how to make it more clear. It reads pretty clear to me but I'll think about it.

File: go/vt/vtctl/workflow/traffic_switcher_test.go
Comment (deepthi) [reply go/vt/vtctl/workflow/traffic_switcher_test.go:117]: oh you mean if you call it `...RequestResponse`? I think `expectApplySchemaResponse` is adequate.

File: proto/vtctldata.proto
Comment (deepthi) [reply proto/vtctldata.proto:213]: Let's change it then.

File: proto/vtctldata.proto
Comment (deepthi) [reply proto/vtctldata.proto:223]: then changing bool to enum is theoretically a breaking change. BUT, since this is between vtctldclient and vtctld, we can assume that their versions should match and not worry about reserving the field.

File: go/cmd/vtctldclient/command/vreplication/movetables/movetables.go
Comment (deepthi) [reply go/cmd/vtctldclient/command/vreplication/movetables/movetables.go:59]: As long as we are doing it in any one place, it's fine with me. I'll leave it to you to decide whether to fail-fast in `create`.

File: proto/vtctldata.proto
Comment (mattlord) [reply proto/vtctldata.proto:223]: The potential problem with renaming the proto field is that the field name is used when marshalling/unmarshalling the field to protojson or prototext. So it would mean that we do not read the value properly for old/existing workflows ... which is fine because in the older versions any actions taken from that value are done when creating the workflow and the value is not needed/used after that.

File: proto/vtctldata.proto
Comment (mattlord) [reply proto/vtctldata.proto:213]: Now I remember that this is what I had started with, but you cannot:
```
vtctldata.proto:213:3: "NONE" is already defined in "vtctldata".
```

See: https://github.com/protocolbuffers/protobuf/issues/5425

So we can leave it at LEAVE, or change it to e.g. NO_HANDLING etc.

File: go/cmd/vtctldclient/command/vreplication/movetables/movetables.go
Comment (mattlord) [reply go/cmd/vtctldclient/command/vreplication/movetables/movetables.go]: You're probably replying to my initial comment. I later edited it to note that we are already checking it in `create` as well as when we do to actually use it.

File: proto/vtctldata.proto
Comment (mattlord) [reply proto/vtctldata.proto]: I tried to address this here: https://github.com/vitessio/vitess/pull/16860/commits/733198b0c7bbc7368a5ce007b28a87890fbb0f2e

File: go/vt/vtctl/workflow/traffic_switcher_test.go
Comment (mattlord) [reply go/vt/vtctl/workflow/traffic_switcher_test.go]: I changed it here: https://github.com/vitessio/vitess/pull/16860/commits/733198b0c7bbc7368a5ce007b28a87890fbb0f2e

File: go/cmd/vtctldclient/command/vreplication/common/utils.go
Comment (mattlord) [reply go/cmd/vtctldclient/command/vreplication/common/utils.go]: I tried to address this here: https://github.com/vitessio/vitess/pull/16860/commits/733198b0c7bbc7368a5ce007b28a87890fbb0f2e

The change is subtle, but I think it's quite clear.

File: go/cmd/vtctldclient/command/vreplication/movetables/create.go
Comment (mattlord) [reply go/cmd/vtctldclient/command/vreplication/movetables/create.go]: I tried to address this here: https://github.com/vitessio/vitess/pull/16860/commits/733198b0c7bbc7368a5ce007b28a87890fbb0f2e

**Review by deepthi — APPROVED** [review]

**Comment by deepthi** [comment]
We'll run the website docs update for the vtctld command reference once before release.

File: changelog/21.0/21.0.0/summary.md
Comment (deepthi) [inline changelog/21.0/21.0.0/summary.md:24]: Was adding `</a>` to every line done by your editor? Seems unnecessary.

File: changelog/21.0/21.0.0/summary.md
Comment (deepthi) [inline changelog/21.0/21.0.0/summary.md:214]: This is too verbose for a summary. Can you please condense it into a short paragraph? No need to give all the background, just describe the feature briefly.

File: changelog/21.0/21.0.0/summary.md
Comment (mattlord) [reply changelog/21.0/21.0.0/summary.md:24]: I did it as w/o it the content is messed up in some readers (the preview one I use). It’s proper to end it.

File: changelog/21.0/21.0.0/summary.md
Comment (mattlord) [reply changelog/21.0/21.0.0/summary.md:214]: In my mind if we add something to the summary we think it’s worth highlighting. At a minimum we should then describe what the thing is that we’re highlighting and explain why we think it’s worth highlighting. Which to me includes the context for what it is, what it is, why it’s important and how to use it. Two paragraphs is not a lot IMO and if you look at this and previous summary docs multiple paragraphs is more the norm than not. That being said, I will try and condense it and eliminate some content that isn’t strictly necessary for the stated objectives.

Comment (mattlord) [reply changelog/21.0/21.0.0/summary.md:214]: I created a docs PR for this: https://github.com/vitessio/website/pull/1859

Once that's merged this can become a short paragraph that links to this for additional details: https://deploy-preview-1859--vitess.netlify.app/docs/21.0/reference/vreplication/movetables/#auto-increment-handling

Comment (mattlord) [reply changelog/21.0/21.0.0/summary.md]: Updated in https://github.com/vitessio/vitess/pull/16860/commits/3ceb25362bf7a3b90fb9a24d49b28662e992a493

**Pull request merged**
